src/sentry/utils/appleconnect/itunes_connect.py

```diff
@@ -349,3 +320,396 @@ def get_dsym_url(
+        self.service_key = service_key
+        self.state = ClientState.NEW
+
+        # The x-apple-id-session-id header as populated by :meth:`start_login_sequence`.
+        self._session_id: Optional[str] = None
+
+        # The scnt header as populated by :meth:`start_login_sequence`.
+        self._scnt: Optional[str] = None
+
+        # The trusted phone info, set by :meth:`_request_trusted_phone_info`.
+        self._trusted_phone: Optional[TrustedPhoneInfo] = None
+
+    @classmethod
+    def from_session_cookie(
+        cls, cookie: str, service_key: Optional[ITunesServiceKey] = None
+    ) -> "ITunesClient":
+        """Creates a client from an existing session cookie.
+
+        :raises SessionExpiredError: if the session cookie is no longer valid.
+        """
+        client = cls(service_key)
+        client.load_session_cookie(cookie)
+        if client.state is not ClientState.AUTHENTICATED:
+            raise SessionExpiredError
+        return client
+
+    def request_auth_service_key(self) -> ITunesServiceKey:
+        """Obtains the authentication service key used in X-Apple-Widget-Key header.
+
+        This key is pretty static and can be cached long-term, saving a request to fetch it.
+        It can optionally be provided to :meth:`init`.
+
+        :returns: the service key to be used in all future calls X-Apple-Widget-Key headers.
+
+        :raises: any exception in case of failure.
+        """
+        url = "https://appstoreconnect.apple.com/olympus/v1/app/config?hostname=itunesconnect.apple.com"
+        logger.debug("GET %s", url)
+        response = self.session.get(url)
+        data = response.json()
+        return ITunesServiceKey(data["authServiceKey"])
+
+    def start_login_sequence(self, username: str, password: str) -> None:
+        """Starts the login process.
+
+        This will trigger the two factor authentication code to show on any validated device
+        for the user.  The code must be provided using :meth:`auth_code`.
+
+        Alternatively after calling this you can switch to SMS two factor authentication by
+        calling :meth:`request_sms`.
+        """
+        assert self.state in [ClientState.NEW, ClientState.EXPIRED]
+        url = "https://idmsa.apple.com/appleauth/auth/signin"
+        logger.debug("POST %s", url)
+
+        start_login = self.session.post(
+            url,
+            json={
+                "accountName": username,
+                "password": password,
+                "rememberMe": True,
+            },
+            headers={
+                "X-Requested-With": "XMLHttpRequest",
+                "X-Apple-Widget-Key": self.service_key,
+                "Accept": "application/json, text/javascript",
+            },
+        )
+
+        if start_login.status_code == http.HTTPStatus.OK:
+            raise ITunesError(
+                f"Two factor auth not enabled for user, status_code={start_login.status_code}"
+            )
```

Comment:
why is this a hard error though? Doesn’t that mean we are done authenticating? Or is the API restricted for all users who don’t have 2FA enabled?

Reply:
afaik users have no choice but have 2FA enabled currently and basically this means our expectation for the server-side are out of the window and stuff probably changed.